File: 2021/07.js
// file reading
fs = require('fs')

function p(input, solver) {
    fs.readFile(input, 'utf8', function(error, data) {
        var lines = data.trim().split('\n');
        console.log(input, solver(lines));
    });
}

function fuel(ps, pos) {
    let result = 0;
    for(let p of ps) {
        result += gauss(Math.abs(p - pos));
    }
    return result;
}

function gauss(x) {
    return (x * (x + 1)) / 2;
}

// solvers
function solve1(lines) {
    const ps = lines[0].split(',').map(x => +x);
    let min = Math.min.apply(null, ps);
    let max = Math.max.apply(null, ps);

    let smallest = Number.MAX_VALUE;
    let smallesti = 0;
    for (let i = min; i <= max; i++) {
        let fuelreq = fuel(ps, i);
        if (fuelreq < smallest) {
            smallest = fuelreq;
            smallesti = i;
        }
    }

    return {smallesti, smallest};
}

p('07-input.txt', solve1);
// p('07-input.txt', solve1);

File: 2021/07.cs
using Aoc.Library.Puzzles;

namespace Aoc2021;

public class Day07 : Puzzle
{
	public override int Day => 7;
	
	public override string SolvePartA(string[] input)
	{
		return Solve(input, CalculateFuel);
	}
	
	public override string SolvePartB(string[] input)
	{
		return Solve(input, CalculateGaussianFuel);
	}

	private static string Solve(string[] input, Func<int[], int, int> fuelCalculator)
	{
		var crabs = input[0].Split(',').Select(int.Parse).ToArray();

		var leastFuel = int.MaxValue;
		for (var i = crabs.Min(); i <= crabs.Max(); i++)
		{
			var fuel = fuelCalculator(crabs, i);
			if (fuel < leastFuel)
			{
				leastFuel = fuel;
			}
		}

		return leastFuel.ToString();
	}

	private int CalculateFuel(int[] crabs, int position) =>
		crabs.Aggregate(0, (subtotal, crab) => subtotal + Math.Abs(crab - position));
	
	private int CalculateGaussianFuel(int[] crabs, int position) =>
		crabs.Aggregate(0, (subtotal, crab) => subtotal + Gauss(Math.Abs(crab - position)));

	private int Gauss(int n) => n * (n + 1) / 2;
}